### PR TITLE
hw1 submission

### PR DESCRIPTION
**Pull request by devusexu**

### CODE REVIEW

File: hw1/src/Homework.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
+<?php declare(strict_types=1);
+
+namespace App;
+
+class Homework
```
Comment (JYu1999) [inline hw1/src/Homework.php]: 這個命名有點奇怪

File: hw1/src/Homework.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
+<?php declare(strict_types=1);
+
+namespace App;
+
+class Homework 
+{
+    public function solve(array $nums, int $target): int
```
Comment (JYu1999) [inline hw1/src/Homework.php]: 這個命名也讓人看不懂這個 function 是在幹嘛

File: hw1/tests/HomeworkTest.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+use PHPUnit\Framework\TestCase;
+
+class HomeworkTest extends Testcase 
+{
+    public static function solveProvider(): array 
+    {
+        // use a big array to ensure the execution time wouldn't be too long
+        // $bigArray consists of 9999998 of 0 and 2 of 9999999 at the end(since it should be sorted)
+        $bigArray = array_fill(1, 9999999, 0);
+        $bigArray[9999999] = $bigArray[10000000] = 9999999;
```
Comment (JYu1999) [inline hw1/tests/HomeworkTest.php]: 請問這樣做的用意是什麼？`bigArray[10000000]` 為何要特地讓其等於 9999999？

File: hw1/tests/HomeworkTest.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+    #[DataProvider('solveProvider')]
+    public function testSolve(array $nums, int $target, int $expected): void 
+    {
+        $hw = new Homework();
+
+        // count elapsed time by recording timestamp before and after executing the method
+        $start = microtime(true);
+        $result =  $hw->solve($nums, $target);
+        $time_elapsed_secs = microtime(true) - $start;
```
Comment (JYu1999) [inline hw1/tests/HomeworkTest.php]: 請使用 camel case

File: hw1/tests/HomeworkTest.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        // sub-array: [nums array, target value, expected answer]
+        return [
+            [[-1,0,3,5,9,12], 9, 4],
+            [[-1,0,3,5,9,12], 10, -1],
+            [$bigArray, 9999999, 9999999]
+        ];
+    }
+
+    #[DataProvider('solveProvider')]
+    public function testSolve(array $nums, int $target, int $expected): void
```
Comment (JYu1999) [inline hw1/tests/HomeworkTest.php]: 基本上，測試會是一個 function 測一筆，這樣 code 會比較清楚。
PHPUnit 應該可以做到，你可以研究看看～

File: hw1/tests/HomeworkTest.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        // $bigArray consists of 9999998 of 0 and 2 of 9999999 at the end(since it should be sorted)
+        $bigArray = array_fill(1, 9999999, 0);
+        $bigArray[9999999] = $bigArray[10000000] = 9999999;
+
+        // every sub-array in the provider array is a test case
+        // sub-array: [nums array, target value, expected answer]
+        return [
+            [[-1,0,3,5,9,12], 9, 4],
+            [[-1,0,3,5,9,12], 10, -1],
+            [$bigArray, 9999999, 9999999]
```
Comment (JYu1999) [inline hw1/tests/HomeworkTest.php]: 只有三筆測資嗎？有沒有什麼其他想法？

**Review by JYu1999 — COMMENTED** [review]
Done

File: hw1/tests/HomeworkTest.php (hunk 3 above)
Comment (devusexu) [reply hw1/tests/HomeworkTest.php]: 因為測資要是排序好的陣列，因此讓bigArray[10000000]>=9999999
有一千萬筆則是按照[補充說明](https://github.com/Goodidea-backend-camp/BECamp12-1-formal/tree/main#:~:text=%F0%9F%A4%94%20If%20there%20is%20an%20array%20that%20have%20one%20million%20number%2C%20which%20is%20like%20%5B1%2C%202%2C%203%2C%20...%2C%20100000000%5D.%20You%20should%20ensure%20that%20user%20can%20find%209%2C999%2C999%20within%205ms.)

File: hw1/tests/HomeworkTest.php (hunk 4 above)
Comment (devusexu) [reply hw1/tests/HomeworkTest.php]: 好的

File: hw1/tests/HomeworkTest.php (hunk 3 above)
Comment (JYu1999) [reply hw1/tests/HomeworkTest.php]: 那為什麼 `bigArray[10000000]` 不能等於 10000000，而是要等於 9999999 呢？

Comment (devusexu) [reply hw1/tests/HomeworkTest.php]: 當然可以，只是因為這樣可以寫短一點
不如我等等把它刪掉好了，畢竟多一筆也沒什麼意思

File: hw1/tests/HomeworkTest.php (hunk 5 above)
Comment (devusexu) [reply hw1/tests/HomeworkTest.php]: 這邊我想對同一個 function 測不同資料(包括範例中的兩個 case，再加上一個大 array 作為 edge case)，使用 data provider 提供多筆同樣格式的資料就不用重複寫 assertion，若擔心不知哪筆報錯可以為每筆測資加上 index

Comment (JYu1999) [reply hw1/tests/HomeworkTest.php]: 原來還可以加 index🧐
那你加一下好了～比較方便辨別

File: hw1/tests/HomeworkTest.php (hunk 3 above)
Comment (JYu1999) [reply hw1/tests/HomeworkTest.php]: 你可以思考一下，你這樣設計，在某些情況下會不會有問題產生～

Comment (devusexu) [reply hw1/tests/HomeworkTest.php]: 我只有想到這樣設計有重複的數字，以及在 array_fill 的時候我忘記要從 index 0 開始
我另外生了一些測資

File: hw1/tests/HomeworkTest.php (hunk 6 above)
Comment (devusexu) [reply hw1/tests/HomeworkTest.php]: 我加了幾筆測資，包含陣列裡有重複 target 以及 target 太大太小可以直接 return
我原本還打算考慮 target 超過儲存範圍，但我發現超過範圍就會變成 float，type hint 會報錯不用另外抓

File: hw1/tests/HomeworkTest.php (hunk 5 above)
Comment (devusexu) [reply hw1/tests/HomeworkTest.php]: 加了！

**Comment by devusexu** [comment]
我改了命名、增加測資、修改主要程式碼以回傳第一個符合 target 的 index
不曉得這樣假設對不對，還是應該回傳所有符合的 index 呢？

File: hw1/tests/HomeworkTest.php (hunk 3 above)
Comment (JYu1999) [reply hw1/tests/HomeworkTest.php]: 我也覺得有重複的數字是一個問題，但怎麼解決是另外一個問題XD

File: hw1/src/IntegerFinder.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace App;
+
+class IntegerFinder 
+{
+    public function findIntegerInArray(array $nums, int $target): int 
+    {
+        // since nums is sorted, if the target is out of range, return -1 immediately
+        // I did not use reset() or end() since it moves the pointer, resulting in longer execution
+        if ($target < $nums[array_key_first($nums)] || $target >  $nums[array_key_last($nums)]) {
```
Comment (JYu1999) [inline hw1/src/IntegerFinder.php:11]: 這個判斷可以考慮包成另一個 method

File: hw1/src/IntegerFinder.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
+<?php declare(strict_types=1);
+
+namespace App;
+
+class IntegerFinder 
+{
+    public function findIntegerInArray(array $nums, int $target): int 
+    {
+        // since nums is sorted, if the target is out of range, return -1 immediately
+        // I did not use reset() or end() since it moves the pointer, resulting in longer execution
```
Comment (JYu1999) [inline hw1/src/IntegerFinder.php:10]: 註解不僅是給自己看的，也是要給別人看的。第十行我看起來有點像是你個人的筆記，可以思考一下這行有沒有必要給別人看。

File: hw1/tests/IntegerFinderTest.php [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        $finder = new IntegerFinder();
+
+        // count elapsed time by recording timestamp before and after executing the method
+        $start = microtime(true);
+        $result =  $finder->findIntegerInArray($nums, $target);
+        $timeElapsed = microtime(true) - $start;
+
+        // result should be same as the expected value
+        $this->assertSame($expected, $result);
+        // execution time should be less than 5ms
```
Comment (JYu1999) [inline hw1/tests/IntegerFinderTest.php:38]: 建議註解上方如果不是判斷式的話，可以多換個一行，可讀性比較高

**Review by JYu1999 — COMMENTED** [review]
Done
